### PR TITLE
handle unknown packet type error case gracefully 

### PR DESCRIPTION
**Pull request by g1rana**
This issue was reported in https://github.com/contiv/netplugin/issues/891. Contiv/netplugin process cored when handling connection flow tracking event. On debugging core logs and available logs in the issue, it was found out that libOpenflow package was not handling gracefully packet flow event for the flows which are not supported yet. This fix is handling such errors gracefully, avoiding null access on unsupported flows

### CODE REVIEW

**Comment by dseevr** [comment]
1.  Could you rewrite the title of this commit (`git commit --amend && git push -f`) so that it includes a high-level overview of the changes being made?  Referring to issue numbers is not useful in the first line of a commit message.  It's fine to put additional detail in the body, but the first line of the commit message is what will show up in `git log`, on Github in various places, etc.
1.  If you include "Fixes <url_to_issue>" in the body of the commit message (line 3 or >), the linked issue will be automatically closed when the PR is merged, e.g.,:

```
Fixed a bug where X fails when Y is set

Fixes <github_issue_url>
```

File: openflow13/match.go
Comment (dseevr) [inline openflow13/match.go]: `= nil` is not required here.  `var err error` is sufficient

File: openflow13/match.go
Comment (dseevr) [inline openflow13/match.go]: This comment is not necessary, returning formatted error messages is extremely common in Go

File: openflow13/match.go
Comment (dseevr) [inline openflow13/match.go]: Please remove this comment, too

File: openflow13/openflow13.go
Comment (dseevr) [inline openflow13/openflow13.go]: When an error (or any other value returned from a function) is not used _except_ inside a single branch `if` statement immediately following it, you should declare it inside the `if` statement scope, e.g.,:

```go
if err := p.Match.UnmarshalBinary(data[n:]); err != nil {
        return err
}
```

This pattern is used everywhere in Go codebases.  It's better to shadow an existing variable (the function-level `err`) than to change its value for no reason.

In this function, there shouldn't even be a function-level `err` variable defined.  The `if` statement at the top of the function should also be following this inline assignment pattern.  I see a bunch of other functions which are also assigning function-level errors and not checking them, too, but these are problems in the existing code and can be fixed later. 😢

**Review by dseevr — CHANGES_REQUESTED** [review]

File: openflow13/match.go
Comment (dseevr) [reply openflow13/match.go]: Ah, I see that you copied the existing code... please do not assume that existing code is correct or following Go best practices because in many cases it is not 😄 

You could also delete the `= 0` here.  0 is the default value of all integer types in Go

Comment (g1rana) [reply openflow13/match.go]: Got it ; Yes, Go complier would init all declare  variable with 0 or nil (incase of reference type)

File: openflow13/match.go
Comment (g1rana) [reply openflow13/match.go]: ok; done

File: openflow13/match.go
Comment (g1rana) [reply openflow13/match.go]: Ok

File: openflow13/openflow13.go
Comment (g1rana) [reply openflow13/openflow13.go]: ok

File: :
Comment (dvavili) [inline :]: Looks like a type. Please remove this file.

File: openflow13/match.go
Comment (dvavili) [inline openflow13/match.go]: Can remove this comment as well. It's pretty straightforward that it's a basic sanity check.

File: openflow13/match.go
Comment (dvavili) [inline openflow13/match.go]: Please add error check here.

**Review by dvavili — COMMENTED** [review]
The changes in this PR will fix the issue that resulted in a crash in netplugin on receiving an unknown packet. But there are a lot of cases in other UnmarshalBinary functions that need to be handled for error checking as well and that was the intent of the issue #2. 

Could you also explain on how you tested these changes? If this was tested against an unknown packet and it prevented a crash we can updated netplugin accordingly and close contiv/netplugin#891, but let's keep Issue #2 open for now. Also, please update the title of the PR accordingly as there are more cases to handle packet errors gracefully.

File: openflow13/match.go
Comment (dseevr) [inline openflow13/match.go]: This error check should be inline in the `if` statement because `err` is never used again

File: :
Comment (g1rana) [reply :]: According current code layout , it doesn't look that any  other flow parser except "Type_PacketIn" is doing building match filed TLV processing and  issue is cause because of match rule processing in context of "Type_PacketIn" parser event. As of now, we can consider this issue is isolated to "Type_PacketIn" flows. 

```
        case Type_PacketIn:
                message = new(PacketIn)
                err = message.UnmarshalBinary(b)
```

```
func (p *PacketIn) UnmarshalBinary(data []byte) error {
        err := p.Header.UnmarshalBinary(data)
        n := p.Header.Len()

        p.BufferId = binary.BigEndian.Uint32(data[n:])
        n += 4
        p.TotalLen = binary.BigEndian.Uint16(data[n:])
        n += 2
        p.Reason = data[n]
        n += 1
        p.TableId = data[n]
        n += 1
        p.Cookie = binary.BigEndian.Uint64(data[n:])
        n += 8

        if err := p.Match.UnmarshalBinary(data[n:]); err != nil {
                return err
        }
        n += p.Match.Len() . ====> before fix, core is happening  at this place on referencing nil .
```

File: openflow13/match.go
Comment (g1rana) [reply openflow13/match.go]: sure,

File: openflow13/match.go
Comment (g1rana) [reply openflow13/match.go]: Ok . Will put check .

File: openflow13/match.go
Comment (g1rana) [reply openflow13/match.go]: Ok.  Done

**Review by dseevr — APPROVED** [review]

**Comment by unclejack** [comment]
I'll test this and comment again.

**Review by rhim — APPROVED** [review]
LGTM.

**Comment by rhim** [comment]
@g1rana have you addressed comments from @DivyaVavili ?

**Comment by g1rana** [comment]
@rhim @srampal  , this issue also needs to merge

**Comment by unclejack** [comment]
@g1rana: Please accept my apologies. I haven't had a chance to test yet, but I really need to test this tomorrow.

**Comment by unclejack** [comment]
This seems to be OK. After this PR is merged, a PR to update this in ofnet needs to be sent. Once the PR to bump libOpenflow in ofnet is merged, a PR needs to be sent against netplugin to update ofnet & libOpenflow.

**Comment by dseevr** [comment]
@contiv/committers-core-networking PTAL

**Review by gkvijay — APPROVED** [review]

**Pull request merged**
